warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java

```diff
@@ -0,0 +1,190 @@
+    } else if (classSelector.startsWith("=")) {
+      this.classnamePattern = Pattern.compile(Pattern.quote(classSelector.substring(1))).matcher("");
+    } else {
+      this.classnamePattern = Pattern.compile(Pattern.quote(classSelector)).matcher("");
+    }
+
+    // read label selector
+    String labelsSelection = m.group("labels");
+    Map<String, String> labelsSelectors = null;
+    try {
+      labelsSelectors = GTSHelper.parseLabelsSelectors(labelsSelection);
+    } catch (ParseException pe) {
+      throw new WarpScriptException(pe);
+    }
+
+    if (0 == labelsSelectors.size()) {
+      this.labelsPatterns = null;
+    } else {
+      this.labelsPatterns = new HashMap<>(labelsSelectors.size());
+      // build label patterns map
+      for (Entry<String, String> l: labelsSelectors.entrySet()) {
+        if (l.getValue().startsWith("=")) {
+          this.labelsPatterns.put(l.getKey(), Pattern.compile(Pattern.quote(l.getValue().substring(1))).matcher(""));
+        } else { //starts with ~ , otherwise Parse Exception in GTSHelper.parseLabelsSelectors
+          this.labelsPatterns.put(l.getKey(), Pattern.compile(l.getValue().substring(1)).matcher(""));
+        }
+      }
+    }
+
+    // read attribute selector, if any
+    Map<String, String> attributesSelectors = null;
+    String attributesSelection = m.group("attributes");
+    if (attributesSelection != null) {
+      try {
+        attributesSelectors = GTSHelper.parseLabelsSelectors(attributesSelection);
+      } catch (ParseException pe) {
+        throw new WarpScriptException(pe);
+      }
+      this.attributesPatterns = new HashMap<>(attributesSelectors.size());
+      // build label patterns map
+      for (Entry<String, String> l: attributesSelectors.entrySet()) {
+        if (l.getValue().startsWith("=")) {
+          this.attributesPatterns.put(l.getKey(), Pattern.compile(Pattern.quote(l.getValue().substring(1))).matcher(""));
+        } else { //starts with ~ , otherwise Parse Exception in GTSHelper.parseLabelsSelectors
+          this.attributesPatterns.put(l.getKey(), Pattern.compile(l.getValue().substring(1)).matcher(""));
+        }
+      }
+    } else {
+      this.attributesPatterns = null;
+    }
+  }
+
+  public boolean MetaDataMatch(Metadata metadata) {
+
+    // first, check classname
+    boolean classnameMatch = (null == this.classnamePattern) || null == metadata.getName()
+        || ((null != this.classnamePattern) && this.classnamePattern.reset(metadata.getName()).matches());
+
+    // then, check labels.
+    // standard selector: class{labelOrAttribute=xxx}
+    // if there is no label with this name, look at the attributes.
+    // extended selector: class{label=xxx}{attribute=yyy}
+    // check separately labels and attributes, both must match.
+
+    Map inputLabels = metadata.getLabels();
+    Map inputAttributes = metadata.getAttributes();
+    boolean labelAndAttributeMatch = true;
+
+    if (null != this.attributesPatterns) {
```

Comment:
Make the comment explaining how match is performed a little more verbose

Reply:
The doc and the tests are explicit. I add a few lines from the doc in the code.